ansible_collections/arista/avd/roles/eos_designs/templates/network_services/logic.j2

```diff
@@ -0,0 +1,7 @@
+{% for network_services_key in network_services_keys | arista.avd.natural_sort %}
+{%     if hostvars[inventory_hostname][network_services_key.name] is arista.avd.defined %}
```

Comment:
from tests with large scale inventories, we have learned that using `hostvars` should be avoided, since jinja copies the entire `hostvars` dataset between each call to the underlying code. In our tests it improved the speed a lot if we copy the part we need into a temporary variable.
Ex.
`{% set tmp_network_services = hostvars[inventory_hostname][network_services_key.name] | arista.avd.default({}) %}`
And then reference the temporary variable below.
I think it should be possible to do a single update with `{% do tenants.update(tmp_network_services) %}`

Reply:
This should probably be done for connected_endpoints_keys also then. Since this implementation is pretty close to what connected_endpoints_keys does. Although the improvement might be not that significant as it does not have this second loop.

Reply:
Correct, this was discovered recently, and we will need to refactor all templates where we use potential large data sets.
We plan to change the data models to lists of dicts instead of dicts of dicts. This will give us improvements on speed, since we don't need to address the complete parent dict every time we need a single child key. So all new models will be built with this in mind, and for 4.0 (subject to change :) ) we will try to do the conversion.
The conversion is also part of the work with schema-based input validation and documentation.

Reply:
@melkypie I have implemented the changes to handle future list-of-dicts as well as `network_serivces_keys`. It took a while, but I hope you can test this out, so we can get it merged before our planned release on Friday.